Framework/DataObjects/src/ScanningWorkspaceBuilder.cpp

```diff
@@ -0,0 +1,323 @@
+ *
+ * @param startTime A DateAndTime object corresponding to the start of the first
+ *scan
+ * @param durations A vector of doubles containing the duration in seconds
+ */
+void ScanningWorkspaceBuilder::setTimeRanges(
+    const Kernel::DateAndTime &startTime,
+    const std::vector<double> &durations) {
+  verifyTimeIndexSize(durations.size(), "time durations");
+
+  std::vector<std::pair<Kernel::DateAndTime, Kernel::DateAndTime>> timeRanges =
+      {std::pair<Kernel::DateAndTime, Kernel::DateAndTime>(
+          startTime, startTime + durations[0])};
+
+  for (size_t i = 1; i < m_nTimeIndexes; ++i) {
+    const auto newStartTime = timeRanges[i - 1].second;
+    const auto endTime = newStartTime + durations[i];
+    timeRanges.push_back(std::pair<Kernel::DateAndTime, Kernel::DateAndTime>(
+        newStartTime, endTime));
+  }
+
+  setTimeRanges(timeRanges);
+}
+
+/**
+ * Supply a vector of vectors which contain positions. The inner vectors should
+ *contain the position for each time index, the outer vector the vector for each
+ *detector.
+ *
+ * @param positions A vector of vectors containing positions
+ */
+void ScanningWorkspaceBuilder::setPositions(
+    const std::vector<std::vector<Kernel::V3D>> &positions) {
+
+  if (!m_positions.empty() || !m_instrumentAngles.empty())
+    throw std::logic_error("Can not set positions, as positions or instrument "
+                           "angles have already been set.");
+
+  for (const auto &vector : positions) {
+    verifyTimeIndexSize(vector.size(), "positions");
+  }
+  verifyDetectorSize(positions.size(), "positions");
+
+  m_positions = positions;
+}
+
+/**
+ * Supply a vector of vectors which contain rotations. The inner vectors should
+ *contain the rotation for each time index, the outer vector the vector for each
+ *detector.
+ *
+ * @param rotations A vector of vectors containing rotations
+ */
+void ScanningWorkspaceBuilder::setRotations(
+    const std::vector<std::vector<Kernel::Quat>> &rotations) {
+
+  if (!m_rotations.empty() || !m_instrumentAngles.empty())
+    throw std::logic_error("Can not set rotations, as rotations or instrument "
+                           "angles have already been set.");
+
+  for (const auto &vector : rotations) {
+    verifyTimeIndexSize(vector.size(), "rotations");
+  }
+  verifyDetectorSize(rotations.size(), "rotations");
+
+  m_rotations = rotations;
+}
+
+/**
+ * Set a vector of rotations corresponding to each time index. These angles
+ *rotate the detector banks around the source, setting the corresponding
+ *positions and rotations of the detectors.
+ *
+ * Here explicit assumptions are made - that the source is at (0, 0, 0), and the
+ *rotation is in the X-Z plane. This corresponds to the common case of moving
+ *detectors to increase angular coverage.
+ *
+ * @param instrumentAngles a vector of angles, the size matching the number of
+ *time indexes
+ */
+void ScanningWorkspaceBuilder::setInstrumentAngles(
```

Comment:
I am in two minds about this too. On balance I think this method is general enough to be here, as all real cases for scanning instruments at ILL and MLZ rotate around the sample position in such a way.